### PR TITLE
Dc/taco 151

### PR DESCRIPTION
**Pull request by khappucino**
I Modified ConfigReloader in the following ways:

ConfigReloader is now initialized with a filePath to the top level application.conf. However the ConfigReloader only updates if specific watched files are modified on disk (and they have different content than before).  The interface to add new files to watch is the ConfigReloader.addWatchFile(File file) method. This method will yell at you if the file you specifically tried to watch doesn't exist.  

Every time you call addWatchFile we compute the lastModifiedDate and the file contents digest. This is added to a HashMap keyed off of the file path.  Periodically when the update method is called we iterate through that hashmap of paths and we compare the current version of the watched file metadata (the lastModifiedDate and the computed digest of the contents), if there is a change we will reload the entire config tree and we will also update the metadata of the watch file at that path.

The way that you add an included file in the application config is like this:
application.conf =   trivial {  include "includedFile.conf" }
includedFile.conf = {  limit = "10" }

In the future we can have other included files and the addWatch method can be used to watch changes in those files as well. Any time there is a change in the dependency tree for the top level conf we recompute the entire configuration. 

Interesting findings***    If you provide an absolute path when you call ConfigFactory.parseFile(File file) and your top level file has a reference to another conf file in the same directory, the ConfigFactory.load(File file) method is smart enough to check the other files at the relative path in order to resolve the includedFile.conf file.

However, if you use ConfigFactory.parseReader(Reader reader), the ConfigFactory annoyingly only searches for the includedFile.conf in the existing classpath vs the same relative path as the top level conf.  This is why you see me adding the temp directories to the class path at run time.  This isn't going to be a problem in production because we specify the classpath when we run the nlp/nfe/xio deriatives, however during tests this was not the case.

### CODE REVIEW

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java [hunk 1]
```diff
@@ -72,24 +87,60 @@ private Config parse(Reader reader) {
     return load(new File(file));
   }
 
-  private boolean performUpdate(Meta<T> oldValue, Meta<T> newValue) {
-    return !MessageDigest.isEqual(oldValue.digest, newValue.digest);
+  private boolean performUpdate(ConfigFileMetadata current, ConfigFileMetadata previous) {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: these new parameter names are not as clear (and swapped with previous impl) - `oldValue` / `newValue`

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java [hunk 2]
```diff
@@ -72,24 +87,60 @@ private Config parse(Reader reader) {
...
+    return (current.lastModified > previous.lastModified
+        && !MessageDigest.isEqual(current.digest, previous.digest));
+  }
+
+  private ConfigFileMetadata loadMetaData(File configFile) throws IllegalStateException {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      DigestInputStream digester = new DigestInputStream(new FileInputStream(configFile), digest);
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: i see a lot of `Closable` resources that are not being closed - these InputStreams and Readers should be closed ... consider using `try with resources`

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java [hunk 3]
```diff
@@ -72,24 +87,60 @@ private Config parse(Reader reader) {
...
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(
+          "Illegal digest algorithm used on file '" + configFile.getAbsolutePath() + "'", e);
+    }
+  }
+
+  private boolean haveWatchFilesChanged() {
+    Boolean filesHaveChanged = false;
+    try {
+      Set<String> keys = new HashSet<String>(watchFiles.keySet());
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: "nice to have": java streams api would be "nicer" here

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java [hunk 4]
```diff
@@ -106,6 +157,15 @@ void setUpdater(BiConsumer<T, T> updater) {
     this.updater = updater;
   }
 
+  public void addWatchFile(File file) throws IllegalStateException {
+    if (!watchFiles.containsKey(file.getAbsolutePath())) {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: this if check (whole function rather) isn't really necessary ... this isn't a MultiMap so only one value per key

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 3 above)
Comment (manimaul) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: eg
```java
List<ConfigFileMetadata> toUpdateList = watchFiles.keySet().stream()
      .map(filePath -> loadMetaData(new File(filePath)))
      .filter(md -> {
        ConfigFileMetadata previousConfigMetadata = watchFiles.get(md.path);
        return performUpdate(md, previousConfigMetadata);
      })
      .collect(Collectors.toList());

    toUpdateList.forEach(update -> watchFiles.put(update.path, update));

    return !toUpdateList.isEmpty();
```

File: xio-core/src/test/java/com/xjeffrose/xio/config/ConfigReloaderUnitTest.java [hunk 5]
```diff
@@ -48,122 +65,198 @@ String createConfig(String value) throws IOException, FileNotFoundException {
...
-    assertEquals(10, config.limit);
-    reloader.start(state::update);
-
-    Thread.sleep(2000);
+  private static void addPath(String s) throws Exception {
+    File f = new File(s);
+    URI u = f.toURI();
+    URLClassLoader urlClassLoader = (URLClassLoader) ClassLoader.getSystemClassLoader();
+    Class<URLClassLoader> urlClass = URLClassLoader.class;
+    Method method = urlClass.getDeclaredMethod("addURL", new Class[] {URL.class});
```
Comment (manimaul) [inline xio-core/src/test/java/com/xjeffrose/xio/config/ConfigReloaderUnitTest.java]: line 79 and 81 - array is unnecessary for varargs

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 1 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: ok i can change the name

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 2 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: ok will do

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 3 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: k

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 1 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: name changed.

File: xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java (hunk 2 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/ConfigReloader.java]: swapped to try with resources =p

File: xio-core/src/test/java/com/xjeffrose/xio/config/ConfigReloaderUnitTest.java (hunk 5 above)
Comment (khappucino) [reply xio-core/src/test/java/com/xjeffrose/xio/config/ConfigReloaderUnitTest.java]: ok changed

**Pull request merged**